crates/rpc/rpc/src/admin.rs

```diff
@@ -106,36 +106,70 @@ where
     async fn node_info(&self) -> RpcResult<NodeInfo> {
         let enode = self.network.local_node_record();
         let status = self.network.network_status().await.to_rpc_result()?;
-        let config = ChainConfig {
+        let mut config = ChainConfig {
             chain_id: self.chain_spec.chain.id(),
             terminal_total_difficulty_passed: self
                 .chain_spec
                 .get_final_paris_total_difficulty()
                 .is_some(),
+            terminal_total_difficulty: self
+                .chain_spec
+                .hardforks
+                .fork(EthereumHardfork::Paris)
+                .ttd(),
             ..self.chain_spec.genesis().config.clone()
         };
 
-        let node_info = NodeInfo {
-            id: B256::from_slice(&enode.id.as_slice()[..32]),
-            name: status.client_version,
-            enode: enode.to_string(),
-            enr: self.network.local_enr().to_string(),
-            ip: enode.address,
-            ports: Ports { discovery: enode.udp_port, listener: enode.tcp_port },
-            listen_addr: enode.tcp_addr(),
-            protocols: ProtocolInfo {
-                eth: Some(EthProtocolInfo {
-                    network: status.eth_protocol_info.network,
-                    difficulty: status.eth_protocol_info.difficulty,
-                    genesis: status.eth_protocol_info.genesis,
-                    config,
-                    head: status.eth_protocol_info.head,
-                }),
-                snap: None,
-            },
-        };
+        macro_rules! set_block_or_time {
+            ($config:expr, $field:ident, $fork:ident) => {
+                // don't overwrite if already set
+                if $config.$field.is_none() {
+                    $config.$field = self.chain_spec.hardforks.fork_block(EthereumHardfork::$fork);
+                }
+            };
+        }
 
-        Ok(node_info)
+        set_block_or_time!(config, homestead_block, Homestead);
+        set_block_or_time!(config, dao_fork_block, Dao);
+        set_block_or_time!(config, eip150_block, Tangerine);
+        set_block_or_time!(config, eip155_block, SpuriousDragon);
```

Comment:
this can be further simplified by making the macro accept (,)* so we only have to do

```rust
set_block_or_time!(config, [homestead_block => Homestead, dao_fork_block => Dao,...)])
```

Reply:
coooool